File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    private void dfs(TreeNode root, List<Integer> values){
        if (root == null)
            return;

        dfs(root.left, values);
        values.add(root.val);
        dfs(root.right, values);
    }

    public int kthSmallest(TreeNode root, int k) {
        var values = new ArrayList<Integer>();
        dfs(root, values);
        // System.out.println(values);
        return values.get(k - 1);
    }
}

File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def lowestCommonAncestor(self, root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:
        def dfs(root, k, stack):
            if root:      
                stack.append(root)

                if root.val == k.val:
                    return True

                if dfs(root.left, k, stack) or dfs(root.right, k, stack):
                    return True

                stack.pop()
                return False
            return False
        
        stack1 = []
        stack2 = []

        dfs(root, p, stack1)
        dfs(root, q, stack2)

        point = 0
        while point < len(stack1) and point < len(stack2) and stack1[point] == stack2[point]:
            point += 1
        
        return stack1[point - 1]